staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -235,144 +213,136 @@ const (
+	doneCh := ctx.Done()
+	if doneCh != nil {
+		qs.preCreateOrUnblockGoroutine()
+		go func() {
+			defer runtime.HandleCrash()
+			qs.goroutineDoneOrBlocked()
+			select {
+			case <-doneCh:
+				klog.V(6).Infof("QS(%s): Context of request %#+v %#+v is Done", qs.qCfg.Name, descr1, descr2)
+				qs.cancelWait(req)
 			}
-			req = qs.dispatchSansQueue(descr1, descr2)
-			metrics.UpdateFlowControlRequestsExecuting(qs.qCfg.Name, qs.totRequestsExecuting)
-			return decisionExecute
-		}
+			qs.goroutineDoneOrBlocked()
+		}()
+	}
+	return req, false
+}
 
-		// ========================================================================
-		// Step 1:
-		// 1) Start with shuffle sharding, to pick a queue.
-		// 2) Reject old requests that have been waiting too long
-		// 3) Reject current request if there is not enough concurrency shares and
-		// we are at max queue length
-		// 4) If not rejected, create a request and enqueue
-		req = qs.timeoutOldRequestsAndRejectOrEnqueueLocked(hashValue, descr1, descr2)
-		defer metrics.UpdateFlowControlRequestsInQueue(qs.qCfg.Name, qs.totRequestsWaiting)
-		// req == nil means that the request was rejected - no remaining
-		// concurrency shares and at max queue length already
-		if req == nil {
-			klog.V(5).Infof("QS(%s): rejecting request %#+v %#+v due to queue full", qs.qCfg.Name, descr1, descr2)
-			metrics.AddReject(qs.qCfg.Name, "queue-full")
-			return decisionReject
-		}
+func (req *request) Wait() (bool, bool, func() (idle bool)) {
+	qs := req.qs
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
 
-		// ========================================================================
-		// Step 2:
-		// The next step is to invoke the method that dequeues as much
-		// as possible.
-		// This method runs a loop, as long as there are non-empty
-		// queues and the number currently executing is less than the
-		// assured concurrency value.  The body of the loop uses the
-		// fair queuing technique to pick a queue and dispatch a
-		// request from that queue.
-		qs.dispatchAsMuchAsPossibleLocked()
-		defer metrics.UpdateFlowControlRequestsExecuting(qs.qCfg.Name, qs.totRequestsExecuting)
-
-		// ========================================================================
-		// Step 3:
-
-		// Set up a relay from the context's Done channel to the world
-		// of well-counted goroutines. We Are Told that every
-		// request's context's Done channel gets closed by the time
-		// the request is done being processed.
-		doneCh := ctx.Done()
-		if doneCh != nil {
-			qs.preCreateOrUnblockGoroutine()
-			go func() {
-				defer runtime.HandleCrash()
-				qs.goroutineDoneOrBlocked()
-				select {
-				case <-doneCh:
-					klog.V(6).Infof("QS(%s): Context of request %#+v %#+v is Done", qs.qCfg.Name, descr1, descr2)
-					qs.cancelWait(req)
-				}
-				qs.goroutineDoneOrBlocked()
-			}()
-		}
+	if req.waitStarted {
+		// This can not happen, because the client is forbidden to
+		// call Wait twice on the same request
+		panic(*req)
```

Comment:
Please add a more descriptive note as well, to save time for someone chasing down the panic caused by some inevitable future bug.

Reply:
Are you asking for a more extensive comment, or commentary to go into the argument to `panic`?

Reply:
Commentary for the argument to `panic`. I know someone could follow the stack trace to this line of code and read the comment, but when people are tracking down a panic, any help you can lend is useful (since it's often not just the server that's panicking :wink:)

Reply:
done